File: wbsviz/core/tree_builder.py
"""
Tree builder for constructing WBS tree structure from records.
"""

from typing import List
from wbsviz.core.model import Record, Node, Tree
from wbsviz.core.common import valid_code, parent_code


class TreeBuilder:
    """Builds tree structure from WBS records."""
    
    def __init__(self, strict: bool = True):
        """
        Initialize tree builder.
        
        Args:
            strict: If True, raise error on missing parents.
                   If False, auto-create missing parent nodes.
        """
        self.strict = strict
    
    def build(self, records: List[Record]) -> Tree:
        """
        Build tree from records.
        
        Args:
            records: List of WBS records
            
        Returns:
            Tree structure
            
        Raises:
            ValueError: If codes are invalid or duplicates found (or missing parents in strict mode)
        """
        tree = Tree()
        
        # Create nodes and validate
        for record in records:
            if not valid_code(record.code):
                raise ValueError(f"Invalid code: {record.code}")
            
            if record.code in tree.store:
                raise ValueError(f"Duplicate code: {record.code}")
            
            node = Node()
            node.code = record.code
            node.title = record.title
            node.description = record.description
            node.primaryResp = record.primaryResp
            node.seconderyResp = record.seconderyResp
            node.estimateDuration = record.estimateDuration
            
            tree.store[record.code] = node
        
        # Ensure parents exist (auto-create if !strict)
        def ensure_parent(code: str):
            if not code:
                return
            
            if code not in tree.store:
                if self.strict:
                    raise ValueError(f"Missing parent: {code}")
                
                # Auto-create parent
                parent = Node()
                parent.code = code
                parent.title = f"[Auto] {code}"
                tree.store[code] = parent
                
                # Recursively ensure grandparent
                ensure_parent(parent_code(code))
        
        for code in list(tree.store.keys()):
            ensure_parent(parent_code(code))
        
        # Link children and roots
        for node in tree.store.values():
            parent_code_str = parent_code(node.code)
            if not parent_code_str:
                tree.roots.append(node)
            else:
                parent = tree.store[parent_code_str]
                parent.children.append(node)
        
        # Sort by numeric code segments
        def code_less(a: Node, b: Node) -> bool:
            def split_code(s: str) -> List[int]:
                return [int(x) for x in s.split('.')]
            
            a_parts = split_code(a.code)
            b_parts = split_code(b.code)
            
            min_len = min(len(a_parts), len(b_parts))
            for i in range(min_len):
                if a_parts[i] != b_parts[i]:
                    return a_parts[i] < b_parts[i]
            
            return len(a_parts) < len(b_parts)
        
        # Sort children for each node
        for node in tree.store.values():
            node.children.sort(key=lambda n: tuple(int(x) for x in n.code.split('.')))
        
        # Sort roots
        tree.roots.sort(key=lambda n: tuple(int(x) for x in n.code.split('.')))
        
        return tree

File: legacy/src/TreeBuilder.cpp
#include "wbsviz/TreeBuilder.hpp"
#include "wbsviz/Common.hpp"
#include <algorithm>
#include <sstream>
#include <stdexcept>
#include <functional>

namespace wbsviz {

TreeBuilder::TreeBuilder(bool strict) : strict_(strict) {}

Tree TreeBuilder::build(const std::vector<Record>& recs) {
    Tree t;

    // Create nodes & validate
    for (const auto& r: recs) {
        if (!validCode(r.code)) throw std::runtime_error("Invalid code: " + r.code);
        if (t.store.count(r.code)) throw std::runtime_error("Duplicate code: " + r.code);
        auto n = std::make_unique<Node>();
        n->code = r.code;
        n->title = r.title;
        n->description = r.description;
        n->primaryResp = r.primaryResp;
        n->seconderyResp = r.seconderyResp;
        n->estimateDuration = r.estimateDuration;
        t.store[r.code] = std::move(n);
    }

    // Ensure parents exist (auto-create if !strict_)
    std::function<void(const std::string&)> ensureParent = [&](const std::string& code){
        if (code.empty()) return;
        if (!t.store.count(code)) {
            if (strict_) throw std::runtime_error("Missing parent: " + code);
            auto p = std::make_unique<Node>();
            p->code = code; p->title = "[Auto] " + code;
            t.store[code] = std::move(p);
            ensureParent(parentCode(code));
        }
    };
    for (const auto& kv : t.store) ensureParent(parentCode(kv.first));

    // Link children & roots
    for (auto& kv : t.store) {
        Node* n = kv.second.get();
        auto p = parentCode(n->code);
        if (p.empty()) t.roots.push_back(n);
        else t.store[p]->children.push_back(n);
    }

    // Sort by numeric code segments
    auto codeLess = [](const Node* a, const Node* b){
        auto split = [](const std::string& s){ std::vector<int> v; std::stringstream ss(s); std::string tok; while(std::getline(ss,tok,'.')) v.push_back(std::stoi(tok)); return v; };
        auto va=split(a->code), vb=split(b->code);
        size_t N=std::min(va.size(),vb.size());
        for(size_t i=0;i<N;++i) if (va[i]!=vb[i]) return va[i]<vb[i];
        return va.size()<vb.size();
    };
    for (auto& kv : t.store) std::sort(kv.second->children.begin(), kv.second->children.end(), codeLess);
    std::sort(t.roots.begin(), t.roots.end(), codeLess);

    return t;
}

} // namespace wbsviz
